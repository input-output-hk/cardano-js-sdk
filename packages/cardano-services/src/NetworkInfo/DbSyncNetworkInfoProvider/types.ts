import { Cardano } from '@cardano-sdk/core';

export interface CirculatingSupplyModel {
  circulating_supply: string;
}

export interface TotalSupplyModel {
  total_supply: string;
}

export interface ActiveStakeModel {
  active_stake: string;
}

export interface LiveStakeModel {
  live_stake: string;
}

export interface EpochModel {
  no: number;
}

export interface CostModelsParamModel {
  PlutusV1?: Cardano.CostModel;
  PlutusV2?: Cardano.CostModel;
  PlutusV3?: Cardano.CostModel;
}

export interface ProtocolParamsModel {
  coins_per_utxo_size: string;
  max_tx_size: number;
  max_val_size: string;
  min_pool_cost: string;
  key_deposit: string;
  pool_deposit: string;
  protocol_major: number;
  protocol_minor: number;
  min_fee_a: number;
  min_fee_b: number;
  max_collateral_inputs: number;
  max_block_size: number;
  max_bh_size: number;
  optimal_pool_count: number;
  influence: number;
  monetary_expand_rate: number;
  treasury_growth_rate: number;
  decentralisation: number;
  collateral_percent: number;
  price_mem: number;
  price_step: number;
  max_tx_ex_mem: string;
  max_tx_ex_steps: string;
  max_block_ex_mem: string;
  max_block_ex_steps: string;
  max_epoch: number;
  costs: CostModelsParamModel | null;
  pool_voting_thresholds: Cardano.PoolVotingThresholds;
  drep_voting_thresholds: Cardano.DelegateRepresentativeThresholds;
  min_committee_size: number;
  committee_term_limit: number;
  governance_action_validity_period: number;
  governance_action_deposit: number;
  drep_deposit: number;
  drep_inactivity_period: number;
}
