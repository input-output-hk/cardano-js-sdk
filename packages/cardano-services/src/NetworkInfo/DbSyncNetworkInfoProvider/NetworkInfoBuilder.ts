import {
  ActiveStakeModel,
  CirculatingSupplyModel,
  EpochModel,
  LedgerTipModel,
  TotalSupplyModel,
  WalletProtocolParamsModel
} from './types';
import { Cardano } from '@cardano-sdk/core';
import { Logger, dummyLogger } from 'ts-log';
import { Pool, QueryResult } from 'pg';
import Queries from './queries';

export class NetworkInfoBuilder {
  #db: Pool;
  #logger: Logger;
  constructor(db: Pool, logger = dummyLogger) {
    this.#db = db;
    this.#logger = logger;
  }

  public async queryCirculatingSupply() {
    this.#logger.debug('About to query circulation supply');
    const result: QueryResult<CirculatingSupplyModel> = await this.#db.query(Queries.findCirculatingSupply);
    return result.rows[0].circulating_supply;
  }

  public async queryTotalSupply(maxLovelaceSupply: Cardano.Lovelace) {
    this.#logger.debug('About to query total supply');
    const result: QueryResult<TotalSupplyModel> = await this.#db.query(Queries.findTotalSupply, [maxLovelaceSupply]);
    return result.rows[0].total_supply;
  }

  public async queryActiveStake() {
    this.#logger.debug('About to query active stake');
    const result: QueryResult<ActiveStakeModel> = await this.#db.query(Queries.findActiveStake);
    return result.rows[0].active_stake;
  }

  public async queryLatestEpoch() {
    this.#logger.debug('About to query the last epoch');
    const result: QueryResult<EpochModel> = await this.#db.query(Queries.findLatestCompleteEpoch);
    return result.rows[0].no;
  }

  public async queryLedgerTip() {
    this.#logger.debug('About to query the ledger tip');
    const result: QueryResult<LedgerTipModel> = await this.#db.query(Queries.findLedgerTip);
    return result.rows[0];
  }

  public async queryCurrentWalletProtocolParams() {
    this.#logger.debug('About to query current wallet protocol params');
    const result: QueryResult<WalletProtocolParamsModel> = await this.#db.query(
      Queries.findCurrentWalletProtocolParams
    );
    return result.rows[0];
  }
}
