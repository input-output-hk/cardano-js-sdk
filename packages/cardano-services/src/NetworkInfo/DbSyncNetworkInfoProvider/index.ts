export * from './DbSyncNetworkInfoProvider';
export * as NetworkInfoCacheKey from './keys';
