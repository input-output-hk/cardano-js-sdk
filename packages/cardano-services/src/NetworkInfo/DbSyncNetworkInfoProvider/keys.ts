export const TOTAL_SUPPLY = 'NetworkInfo_total_supply';
export const CIRCULATING_SUPPLY = 'NetworkInfo_circulating_supply';
export const ACTIVE_STAKE = 'NetworkInfo_active_stake';
export const LIVE_STAKE = 'NetworkInfo_live_stake';
export const ERA_SUMMARIES = 'NetworkInfo_era_summaries';
export const LEDGER_TIP = 'NetworkInfo_ledger_tip';
