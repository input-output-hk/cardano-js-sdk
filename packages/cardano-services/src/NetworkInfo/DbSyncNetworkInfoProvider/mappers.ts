import { Cardano, Seconds, SupplySummary } from '@cardano-sdk/core';
import { CostModelsParamModel, ProtocolParamsModel } from './types';
import { GenesisData } from '../../types';
import { LedgerTipModel } from '../../util/DbSyncProvider';

interface ToLovalaceSupplyInput {
  circulatingSupply: string;
  totalSupply: string;
}

export const networkIdMap = {
  Mainnet: Cardano.NetworkId.Mainnet,
  Testnet: Cardano.NetworkId.Testnet
};

export const toSupply = ({ circulatingSupply, totalSupply }: ToLovalaceSupplyInput): SupplySummary => ({
  circulating: BigInt(circulatingSupply),
  total: BigInt(totalSupply)
});

export const toLedgerTip = ({ block_no, slot_no, hash }: LedgerTipModel): Cardano.Tip => ({
  blockNo: Cardano.BlockNo(Number(block_no)),
  hash: hash.toString('hex') as unknown as Cardano.BlockId,
  slot: Cardano.Slot(Number(slot_no))
});

export const mapCostModels = (costs: CostModelsParamModel | null) => {
  const models: Cardano.CostModels = [];
  if (costs?.PlutusV1) models[Cardano.PlutusLanguageVersion.V1] = costs.PlutusV1;
  if (costs?.PlutusV2) models[Cardano.PlutusLanguageVersion.V2] = costs.PlutusV2;
  return models;
};

export const toProtocolParams = ({
  coins_per_utxo_size,
  max_tx_size,
  max_val_size,
  max_collateral_inputs,
  min_pool_cost,
  pool_deposit,
  key_deposit,
  protocol_major,
  protocol_minor,
  min_fee_a,
  min_fee_b,
  max_block_size,
  max_bh_size,
  optimal_pool_count,
  influence,
  monetary_expand_rate,
  treasury_growth_rate,
  decentralisation,
  collateral_percent,
  price_mem,
  price_step,
  max_tx_ex_mem,
  max_tx_ex_steps,
  max_block_ex_mem,
  max_block_ex_steps,
  max_epoch,
  costs
}: ProtocolParamsModel): Cardano.ProtocolParameters => ({
  coinsPerUtxoByte: Number(coins_per_utxo_size),
  collateralPercentage: collateral_percent,
  costModels: mapCostModels(costs),
  decentralizationParameter: String(decentralisation),
  desiredNumberOfPools: optimal_pool_count,
  maxBlockBodySize: max_block_size,
  maxBlockHeaderSize: max_bh_size,
  maxCollateralInputs: max_collateral_inputs,
  maxExecutionUnitsPerBlock: {
    memory: Number(max_block_ex_mem),
    steps: Number(max_block_ex_steps)
  },
  maxExecutionUnitsPerTransaction: {
    memory: Number(max_tx_ex_mem),
    steps: Number(max_tx_ex_steps)
  },
  maxTxSize: max_tx_size,
  maxValueSize: Number(max_val_size),
  minFeeCoefficient: min_fee_a,
  minFeeConstant: min_fee_b,
  minPoolCost: Number(min_pool_cost),
  monetaryExpansion: String(monetary_expand_rate),
  poolDeposit: Number(pool_deposit),
  poolInfluence: String(influence),
  poolRetirementEpochBound: max_epoch,
  prices: {
    memory: price_mem,
    steps: price_step
  },
  protocolVersion: {
    major: protocol_major,
    minor: protocol_minor
  },
  stakeKeyDeposit: Number(key_deposit),
  treasuryExpansion: String(treasury_growth_rate)
});

export const toGenesisParams = (genesis: GenesisData): Cardano.CompactGenesis => ({
  ...genesis,
  networkId: networkIdMap[genesis.networkId],
  slotLength: Seconds(genesis.slotLength),
  systemStart: new Date(genesis.systemStart)
});
