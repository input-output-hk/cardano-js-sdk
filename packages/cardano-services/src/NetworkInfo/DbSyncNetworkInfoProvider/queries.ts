export const findCirculatingSupply = `
    WITH total_rewards AS (
        SELECT COALESCE(SUM(amount), 0) as rewards_amount
        FROM reward
    ),
    total_withdrawals AS (
        SELECT COALESCE(SUM(amount), 0) as withdrawals_amount
        FROM withdrawal
    ),
    total_utxo AS (
        SELECT
            SUM(tx_out.value) AS utxo_amount
        FROM tx_out
        LEFT JOIN tx_in 
            ON tx_out.tx_id = tx_in.tx_out_id
            AND tx_out.index = tx_in.tx_out_index
        WHERE tx_in.id IS NULL    
    )
    SELECT CAST(utxo_amount + (rewards_amount - withdrawals_amount) AS BIGINT) as circulating_supply
    FROM total_rewards, total_withdrawals, total_utxo
`;

export const findTotalSupply = `
    SELECT CAST($1 - reserves AS BIGINT) AS total_supply
    FROM ada_pots
    ORDER BY ada_pots.block_id DESC 
    LIMIT 1
`;

// Active stake is the stake snapshot from n-1 epochs, where n = current epoch
// epoch_stake contains records of completed epochs
export const findActiveStake = `
    SELECT CAST(COALESCE(SUM(amount), 0) AS BIGINT) as active_stake
    FROM epoch_stake
    WHERE epoch_stake.epoch_no = (
    SELECT no FROM epoch
        ORDER BY no DESC
        LIMIT 1
    )
`;

export const findLatestCompleteEpoch = `
    SELECT no
    FROM public.epoch
    ORDER BY no DESC
    LIMIT 1
`;

export const findLedgerTip = `
    SELECT block_no, slot_no, hash
    FROM block
    ORDER BY id DESC
    LIMIT 1;
`;

export const findProtocolParams = `
    SELECT 
    min_fee_a, 
    min_fee_b, 
    max_tx_size,
    key_deposit, 
    pool_deposit, 
    protocol_major, 
    protocol_minor, 
    min_pool_cost, 
    coins_per_utxo_size, 
    max_val_size, 
    max_collateral_inputs,
    max_block_size,
    max_bh_size,
    optimal_pool_count,
    influence,
    monetary_expand_rate,
    treasury_growth_rate,
    decentralisation,
    collateral_percent,
    price_mem,
    price_step,
    max_tx_ex_mem,
    max_tx_ex_steps,
    max_block_ex_mem,
    max_block_ex_steps,
    max_epoch,
    cost_model.costs
    FROM epoch_param
    LEFT JOIN cost_model
        ON cost_model.id = epoch_param.cost_model_id
    ORDER BY epoch_no DESC NULLS LAST
    LIMIT 1;
`;

const Queries = {
  findActiveStake,
  findCirculatingSupply,
  findLatestCompleteEpoch,
  findLedgerTip,
  findProtocolParams,
  findTotalSupply
};

export default Queries;
