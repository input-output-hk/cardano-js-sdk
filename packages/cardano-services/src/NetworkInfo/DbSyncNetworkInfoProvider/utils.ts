import { AsyncAction, InMemoryCache, UNLIMITED_CACHE_TTL } from '../../InMemoryCache';
import { NetworkInfoCacheKey } from '.';
import { Shutdown } from '@cardano-sdk/util';

export const EPOCH_POLL_INTERVAL_DEFAULT = 10_000;

export const epochPollService = (
  cache: InMemoryCache,
  asyncAction: AsyncAction<number>,
  interval: number
): Shutdown => {
  const executePoll = async () => {
    const lastEpoch = await asyncAction();
    const currentEpoch = cache.getVal<number>(NetworkInfoCacheKey.CURRENT_EPOCH);
    const shouldInvalidateEpochValues = !!(currentEpoch && lastEpoch > currentEpoch);

    if (!currentEpoch || shouldInvalidateEpochValues) {
      cache.set<number>(NetworkInfoCacheKey.CURRENT_EPOCH, lastEpoch, UNLIMITED_CACHE_TTL);
      shouldInvalidateEpochValues
        ? cache.invalidate([
            NetworkInfoCacheKey.TOTAL_SUPPLY,
            NetworkInfoCacheKey.ACTIVE_STAKE,
            NetworkInfoCacheKey.ERA_SUMMARIES
          ])
        : void 0;
    }
  };

  const timeout = setInterval(executePoll, interval);

  return {
    shutdown: () => clearInterval(timeout)
  };
};
