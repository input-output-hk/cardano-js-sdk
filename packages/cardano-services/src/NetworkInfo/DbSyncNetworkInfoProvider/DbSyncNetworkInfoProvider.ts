import {
  Cardano,
  CardanoNode,
  CardanoNodeUtil,
  EraSummary,
  NetworkInfoProvider,
  StakeSummary,
  SupplySummary
} from '@cardano-sdk/core';
import { DbSyncProvider } from '../../util/DbSyncProvider';
import { Disposer, EpochMonitor } from '../../util/polling/types';
import { GenesisData } from './types';
import { InMemoryCache, UNLIMITED_CACHE_TTL } from '../../InMemoryCache';
import { Logger } from 'ts-log';
import { NetworkInfoBuilder } from './NetworkInfoBuilder';
import { NetworkInfoCacheKey } from '.';
import { Pool } from 'pg';
import { RunnableModule } from '@cardano-sdk/util';
import { loadGenesisData, toGenesisParams, toLedgerTip, toProtocolParams, toSupply } from './mappers';

export interface NetworkInfoProviderProps {
  cardanoNodeConfigPath: string;
}
export interface NetworkInfoProviderDependencies {
  db: Pool;
  cache: InMemoryCache;
  logger: Logger;
  cardanoNode: CardanoNode;
  epochMonitor: EpochMonitor;
}
export class DbSyncNetworkInfoProvider extends DbSyncProvider(RunnableModule) implements NetworkInfoProvider {
  #logger: Logger;
  #cache: InMemoryCache;
  #builder: NetworkInfoBuilder;
  #genesisDataReady: Promise<GenesisData>;
  #epochMonitor: EpochMonitor;
  #epochRolloverDisposer: Disposer;

  constructor(
    { cardanoNodeConfigPath }: NetworkInfoProviderProps,
    { db, cache, logger, cardanoNode, epochMonitor }: NetworkInfoProviderDependencies
  ) {
    super(db, cardanoNode, 'DbSyncNetworkInfoProvider', logger);
    this.#logger = logger;
    this.#cache = cache;
    this.#epochMonitor = epochMonitor;
    this.#builder = new NetworkInfoBuilder(db, logger);
    this.#genesisDataReady = loadGenesisData(cardanoNodeConfigPath);
  }

  public async ledgerTip(): Promise<Cardano.Tip> {
    const tip = await this.#builder.queryLedgerTip();
    return toLedgerTip(tip);
  }

  public async protocolParameters(): Promise<Cardano.ProtocolParameters> {
    const currentProtocolParams = await this.#builder.queryProtocolParams();
    return toProtocolParams(currentProtocolParams);
  }

  public async genesisParameters(): Promise<Cardano.CompactGenesis> {
    const genesisData = await this.#genesisDataReady;
    return toGenesisParams(genesisData);
  }

  public async lovelaceSupply(): Promise<SupplySummary> {
    const { maxLovelaceSupply } = await this.#genesisDataReady;

    const [circulatingSupply, totalSupply] = await Promise.all([
      this.#cache.get(NetworkInfoCacheKey.CIRCULATING_SUPPLY, () => this.#builder.queryCirculatingSupply()),
      this.#cache.get(
        NetworkInfoCacheKey.TOTAL_SUPPLY,
        () => this.#builder.queryTotalSupply(maxLovelaceSupply),
        UNLIMITED_CACHE_TTL
      )
    ]);

    return toSupply({ circulatingSupply, totalSupply });
  }

  public async stake(): Promise<StakeSummary> {
    this.#logger.debug('About to query stake data');

    const [live, activeStake] = await Promise.all([
      this.#cache.get(NetworkInfoCacheKey.LIVE_STAKE, () =>
        this.cardanoNode.stakeDistribution().then(CardanoNodeUtil.toLiveStake)
      ),
      this.#cache.get(NetworkInfoCacheKey.ACTIVE_STAKE, () => this.#builder.queryActiveStake(), UNLIMITED_CACHE_TTL)
    ]);

    return {
      active: BigInt(activeStake),
      live
    };
  }

  public async eraSummaries(): Promise<EraSummary[]> {
    return await this.#cache.get(
      NetworkInfoCacheKey.ERA_SUMMARIES,
      () => this.cardanoNode.eraSummaries(),
      UNLIMITED_CACHE_TTL
    );
  }

  async initializeImpl() {
    return Promise.resolve();
  }

  async startImpl() {
    this.#epochRolloverDisposer = this.#epochMonitor.onEpochRollover(() => this.#cache.clear());
  }

  async shutdownImpl() {
    this.#cache.shutdown();
    await this.#genesisDataReady;
    this.#epochRolloverDisposer();
  }
}
