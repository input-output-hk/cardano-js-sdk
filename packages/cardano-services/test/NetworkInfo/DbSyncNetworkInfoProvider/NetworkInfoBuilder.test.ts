import { NetworkInfoBuilder } from '../../../src/NetworkInfo/DbSyncNetworkInfoProvider/NetworkInfoBuilder';
import { Pool } from 'pg';

describe('NetworkInfoBuilder', () => {
  let dbConnection: Pool;
  let builder: NetworkInfoBuilder;

  beforeAll(async () => {
    dbConnection = new Pool({ connectionString: process.env.POSTGRES_CONNECTION_STRING });
    builder = new NetworkInfoBuilder(dbConnection);
  });

  afterAll(async () => {
    await dbConnection.end();
  });

  describe('queryCirculatingSupply', () => {
    test('query circulating supply', async () => {
      const result = await builder.queryCirculatingSupply();
      expect(BigInt(result)).toBeGreaterThan(0n);
      expect(result).toMatchSnapshot();
    });
  });

  describe('queryTotalSupply', () => {
    test('query total supply', async () => {
      const maxSupply = 45_000_000_000_000_000n;
      const result = await builder.queryTotalSupply(maxSupply);
      expect(BigInt(result)).toBeGreaterThan(0n);
      expect(result).toMatchSnapshot();
    });
  });

  describe('queryActiveStake', () => {
    test('query active stake', async () => {
      const result = await builder.queryActiveStake();
      expect(BigInt(result)).toBeGreaterThan(0n);
      expect(result).toMatchSnapshot();
    });
  });

  describe('queryLatestEpoch', () => {
    test('query latest epoch', async () => {
      const result = await builder.queryLatestEpoch();
      expect(result).toBeGreaterThan(0);
      expect(result).toMatchSnapshot();
    });
  });

  describe('queryLedgerTip', () => {
    test('query ledger tip', async () => {
      const result = await builder.queryLedgerTip();
      expect({ ...result, hash: result.hash.toString('hex') }).toMatchSnapshot();
    });
  });

  describe('queryCurrentWalletProtocolParams', () => {
    test('query wallet protocol params from current epoch', async () => {
      const result = await builder.queryCurrentWalletProtocolParams();
      expect(result).toMatchSnapshot();
    });
  });
});
