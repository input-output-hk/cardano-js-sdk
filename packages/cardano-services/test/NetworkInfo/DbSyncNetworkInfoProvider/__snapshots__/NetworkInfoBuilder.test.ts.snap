// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NetworkInfoBuilder queryActiveStake query active stake 1`] = `"15039823438862518"`;

exports[`NetworkInfoBuilder queryCirculatingSupply query circulating supply 1`] = `"42022749185221758"`;

exports[`NetworkInfoBuilder queryCurrentWalletProtocolParams query wallet protocol params from current epoch 1`] = `
Object {
  "coins_per_utxo_size": "34482",
  "collateral_percent": 150,
  "decentralisation": 0,
  "influence": 0.3,
  "key_deposit": "2000000",
  "max_bh_size": 1100,
  "max_block_ex_mem": "80000000",
  "max_block_ex_steps": "40000000000",
  "max_block_size": 98304,
  "max_collateral_inputs": 3,
  "max_epoch": 18,
  "max_tx_ex_mem": "16000000",
  "max_tx_ex_steps": "10000000000",
  "max_tx_size": 16384,
  "max_val_size": "5000",
  "min_fee_a": 44,
  "min_fee_b": 155381,
  "min_pool_cost": "340000000",
  "monetary_expand_rate": 0.003,
  "optimal_pool_count": 500,
  "pool_deposit": "500000000",
  "price_mem": 0.0577,
  "price_step": 0.0000721,
  "protocol_major": 6,
  "protocol_minor": 0,
  "treasury_growth_rate": 0.2,
}
`;

exports[`NetworkInfoBuilder queryLatestEpoch query latest epoch 1`] = `205`;

exports[`NetworkInfoBuilder queryLedgerTip query ledger tip 1`] = `
Object {
  "block_no": 3556390,
  "hash": "332340bfcf47951b4bc8eca51c4e9190d29e2fd6dae30be231ebcdadb2d8c399",
  "slot_no": "58336266",
}
`;

exports[`NetworkInfoBuilder queryTotalSupply query total supply 1`] = `"40408479600434778"`;
